enclave_wrangler/dataset_download_new_api.py

```diff
@@ -12,67 +17,175 @@
+         ConceptSuccessorRelationship, ConceptUsageCounts, CsetVersionInfo, CodeSystemConceptSetVersionExpressionItem,
+         OMOPConcept, OMOPConceptAncestorRelationship, OMOPConceptChangeConnectors, OMOPConceptSet,
+         OMOPConceptSetContainer, OMOPConceptSetReview, OmopConceptChange, OmopConceptDomain, OmopConceptSetVersionItem,
+        """
+        url = f'{self.base_url}/api/v1/ontologies/{self.ontology_rid}/objectTypes'
+        response = requests.get(url, headers=self.headers)
+        response_json = response.json()['data']  # always returns everything in nested 'data' key
+        # types = pd.DataFrame(data=response_json)
+        # types = sorted([x['apiName'] for x in response_json])
+        return response_json['data']
+
+    # TODO: Implement this func to get all concept sets
+    #  - handle pagination. I think get 1k at a time
+    # TODO: after: for each set, get all concepts.
+    # TODO: Look at Siggie's frontend/ code to see how OMOPConceptSet etc are being called/used
+    # TODO: Need to find the right object_type, then write a wrapper func around this to get concept sets
+    #  - To Try: CodeSystemConceptSetVersionExpressionItem, OMOPConcept, OMOPConceptSet, OMOPConceptSetContainer,
+    #    OmopConceptSetVersionItem
+    def objects(self, object_type) -> pd.DataFrame:
+        """Get objects
+        Docs: https://www.palantir.com/docs/foundry/api/ontology-resources/objects/list-objects/
+        https://www.palantir.com/docs/foundry/api/ontology-resources/objects/object-basics/"""
+        # Request
+        results: List[Dict] = []
+        url_base = f'{self.base_url}/api/v1/ontologies/{self.ontology_rid}/objects/{object_type}'
+        url = url_base
+        while True:
+            response = requests.get(url, headers=self.headers)
+            response_json = response.json()
+            if response.status_code >= 400:  # err
+                break
+            results += response_json['data']
+            if 'nextPageToken' not in response_json or not response_json['nextPageToken']:
+                break
+            url = url_base + '?pageToken=' + response_json["nextPageToken"]
+
+        # Parse
+        # Get rid of nested 'properties' key, and add 'rid' in with the other fields
+        if results:
+            results = [{**x['properties'], **{'rid': x['rid']}} for x in results]
+            df = pd.DataFrame(results).fillna('')
+
+            # Cache
+            # TODO: temporary cache code until decide where/how to write these
+            cache_dir = os.path.join(self.cache_dir, 'objects', object_type)
+            date_str = str(datetime.now()).replace(':', '-')
+            cache_path = os.path.join(cache_dir, date_str + '.csv')
+            if not os.path.exists(cache_dir):
+                os.mkdir(cache_dir)
+            df.to_csv(cache_path, index=False)
+            if response.status_code >= 400:
```

Comment:
@Sigfried Got an error 500 on OMOPConcept, after it succeeded in returning some 51,000 results.

Reply:
Mmm, no, I didn't do that. But that's not a bad idea; to save the URL of the last failed request (which will include the `pageToken` within it).

_Edit: done:_
```py
            if response.status_code >= 400:
                error_report: Dict = {'request': url, 'response': response_json}
                with open(os.path.join(cache_dir, f'{date_str} - error {response.status_code}.json'), 'w') as file:
                    json.dump(error_report, file)
```